README.md
`version 0.2.4`

## Documentation

Creates SCORM package from source directory.

* Supported Scorm versions:
    * SCORM 1.2
    * SCORM 2004 3rd Edition
    * SCORM 2004 4th Edition

## Installation

```bash
npm install simple-scorm-packager
```

## Initialization Options

* `version` {String} Version of schema. Available options:
    * '1.2'
    * '2004 3rd Edition' alias '2004.3' alias '2004v3'
    * '2004 4th Edition' alias '2004.4' alias '2004v4'
* `organization` {String} Company name
* `language` {String} Language of the package ( ISO )
* `title` {String}
* `identifier` {String} Uses 0 and course title if left empty
* `masteryScore` {Number} Uses 80 if left empty
* `startingPage` {String} Uses index.html if left empty
* `source` {String} The path to files from which the package will be created
* `package` {Object} Available options:
    * `zip` {Boolean} Archives package (`NAME_VERSION_DATE.zip`), false by default
    * `outputFolder` {String} The folder path where you want the zip file
    * `size` {Bytes} Provide the package size, automatically calculated when not set,
    * `name` {String} Package name, defaults to scorm title
    * `useNameAsIs` {Boolean} If true, the `package.name` will be used "as is" without any trimming etc.
    * `author` {String} Author name, used as default for vcard if not provided
    * `version` {String} Package version (major.minor.patch), defaults to `1.0.0`
    * `date` {String} Package date, defaults to now date(YYYY-MM-DD)
    * `vcard` {Object} :
      * `author` {String} Author name, when not provided defaults to package.author
      * `org` {String} Organization name, defaults to `organization`
      * `tel` {String} Telephone number(s)
      * `address` {String} Address
      * `mail` {String} E-mail contact
      * `url` {String} website url
    * `description` {String} Provide the course description or the Project Information
    * `keywords` {Array} Keywords
    * `duration` {String} The time the media takes to play through, format PT#M#S
    * `typicalDuration` {String} The time it will take for a typical learner to fully experience the program, format PT#M#S
    * `requirements` {Array of Objects of the following structure} : 
      * `type` {String} The type of requirement, eg.: Browser, Os
      * `name` {String} The name of the type of requirement, eg.: Microsoft Internet Explorer
      * `version` {String} The minimum version of the requirement
    * `rights` {String} Copyright informations


## Programatically usage

```javascript
var scopackager = require('simple-scorm-packager');

scopackager({
  version: '2004 4th Edition',
  organization: 'Test Company',
  title: 'Test Course',
  language: 'fr-FR',
  identifier: '00',
  masteryScore: 80,
  startingPage: 'index.html',
  source: './myProjectFolder',
  package: {
    name: "lms-package_0.0.1_2004v4",
    useNameAsIs: true,
    version: "0.0.1",
    zip: true,
    outputFolder: './scormPackages'
  }
}, function(msg){
  console.log(msg);
});
```

## USE IT AS CLI
if installed globally you can use it directly in command line

```bash
$ simple-scorm-packager
or
$ scopackager
```

CHANGELOG.MD
# scopackager changelog

## 0.2.5
* Fix issue where file source would be included in the manifest file href's. 

## 0.2.4
* Allow for flag ”useNameAsIs” to use fully custom zip package names

## 0.2.3

* fixed mastery score to 0

## 0.2.2

* fixed package.json

## 0.2.1

* renamed CHANGELOG.MD to CHANGELOG.md

## 0.2.0

* CLI implementation init
* replaced archiver.bulk (deprecated) in favor of archiver.directory

## 0.1.7

* Basic functionality present, mostly spec-compliant

## 0.1.3

* First release capable of doing anything useful
* Still lots of basic functionality missing

## 0.1.0

* Initial experiment

lib/schemas/config.js
var path = require('path');

var now = new Date();
var timestamp = now.getTime();

var nowDate = function() {
    now.setDate(now.getDate());
    return now.getFullYear() + '-' + ('0' + (now.getMonth() + 1)).slice(-2) + '-' + ('0' + now.getDate()).slice(-2);
}

var stringToArray = function(value){
    if(value instanceof Array){
        return value;
    }else if(typeof value === "string"){
        var arr = [];
        if(value) arr.push(value);
        return arr;
    }else{
        var arrOfVals = [];
        for( var k in value ) {
            if(value[k]) arrOfVals.push( value[k] );
        }
        return arrOfVals;
    }
}

var uuid = function() {
    var d = timestamp;
    var uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
        var r = (d + Math.random()*16)%16 | 0;
        d = Math.floor(d/16);
        return (c=='x' ? r : (r&0x3|0x8)).toString(16);
    });
    return uuid;
};

module.exports = function(obj) {
    if (!obj.package) obj.package = {};
    if (!obj.package.vcard) obj.package.vcard = {};
    return {
        uuid: obj.uuid || uuid(),
        version: obj.version || '1.2',
        language: obj.language || 'fr',
        organization: obj.organization || '',
        title: obj.title || '',
        identifier: obj.identifier,
        masteryScore: obj.masteryScore != null ? obj.masteryScore : 80,
        startingPage: obj.startingPage || 'index.html',
        package: {
            name: obj.package.name || obj.title || '',
            useNameAsIs: obj.package.useNameAsIs,
            author:  obj.package.author || '',
            description: obj.package.description || obj.description || '',
            organization: obj.package.organization || obj.organization || '',
            date: obj.package.date || nowDate(),
            timestamp: timestamp,
            size: obj.package.size || '',
            version: obj.package.version || '1.0.0',
            outputFolder: obj.package.outputFolder || './scorm',
            duration : obj.package.duration || 'PT0H0M0S',
            typicalDuration: obj.package.typicalDuration || obj.package.duration || 'PT0H0M0S',
            educational: obj.package.educational || obj.package.description || obj.description || '',
            rights: obj.package.rights || 'Pourvu que-est; à utiliser uniquement à titre d\'exemple.',
            requirements: obj.package.requirements || [],
            keywords: stringToArray(obj.package.keywords || ''),
            status: obj.package.status || 'final',
            vcard: {
                version: obj.package.vcard.version || obj.package.version || '1.0.0',
                author: obj.package.vcard.author || obj.package.author || '',
                org: obj.package.vcard.organization || obj.package.organization || obj.organization || '',
                tel: stringToArray(obj.package.vcard.tel || ''),
                address: stringToArray(obj.package.vcard.address || ''),
                mail: stringToArray(obj.package.vcard.mail || ''),
                url: obj.package.vcard.url || ''
            },
            zip: obj.package.zip
        },
        source: obj.source ? path.normalize(obj.source) : path.normalize('./')
    };
};

lib/index.js
var fs = require("fs"),
  fse = require("fs-extra"),
  path = require("path"),
  getDirectorySize = require("./directory.js"),
  files = require("./files.js"),
  config = require("./schemas/config"),
  schema = require("./schema.js"),
  metadata = require("./metadata.js"),
  manifest = require("./manifest.js"),
  utils = require("./utils");

var _logSuccess = function(msg) {
  var date = new Date();
  var time = date.getHours() + ":" + date.getMinutes() + ":" + date.getSeconds();
  console.log("[" + time + "]", "SCORM", "'" + "\x1b[32m" + msg + "\x1b[0m" + "'");
};

var _logError = function(err) {
  var date = new Date();
  var time = date.getHours() + ":" + date.getMinutes() + ":" + date.getSeconds();
  console.log("[" + time + "]" + "\x1b[31m", err, "\x1b[0m");
};

var buildPackage = function(obj, callback) {
  _logSuccess("Init");
  callback = callback || _logSuccess;
  obj = config(obj);
  var schemaVersion, schemaDefinition;

  switch (obj.version) {
    case "1.2":
      schemaVersion = "scorm12";
      schemaDefinition = "scorm12edition";
      break;
    case "2004.3":
    case "2004v3":
    case "2004 3rd Edition":
      schemaVersion = "scorm2004";
      schemaDefinition = "scorm20043rdedition";
      break;
    case "2004.4":
    case "2004v4":
    case "2004 4th Edition":
      schemaVersion = "scorm2004";
      schemaDefinition = "scorm20044thedition";
      break;
  }

  if (!schemaVersion) {
    _logError("Supported versions:\n1.2\n2004 3rd Edition\n2004 4th Edition");
    return;
  }

  if (obj.package.size === "") {
    getDirectorySize(path.resolve(obj.source), null, function(err, bytes) {
      if (err) {
        throw err;
      }
      obj.package.size = bytes;
    });
  }

  var rootDir = path.dirname(fs.realpathSync(__filename)),
    definitionFileList = files(path.join(rootDir, "schemas", "definitionFiles", schemaDefinition)).map(function(file) {
      return {
        name: file,
        source: path.join(rootDir, "schemas", "definitionFiles", schemaDefinition, file),
        destination: path.join(obj.source, file),
      };
    });

  fse.outputFile(path.join(obj.source, "imsmanifest.xml"), manifest(schemaVersion, obj), function(err) {
    if (err) {
      return _logError(err);
    } else {
      _logSuccess("create " + path.join(obj.source, "imsmanifest.xml"));
      fse.outputFile(path.join(obj.source, "metadata.xml"), metadata(obj), function(err) {
        if (err) {
          return _logError(err);
        } else {
          _logSuccess("create " + path.join(obj.source, "metadata.xml"));
          definitionFileList.forEach(function(file) {
            try {
              fse.copySync(file.source, file.destination);
              _logSuccess("create " + file.destination);
            } catch (err) {
              _logError(err);
            }
          });
          if (obj.package.zip) {
            var archiver = require("archiver");
            fse.ensureDirSync(obj.package.outputFolder);
            var shouldUseNameAsIs = Boolean(obj.package.useNameAsIs);
            var zipOutput = path.join(
              obj.package.outputFolder,
              shouldUseNameAsIs
                ? obj.package.name + ".zip"
                : utils.cleanAndTrim(obj.package.name) + "_v" + obj.package.version + "_" + obj.package.date + ".zip"
            );
            _logSuccess("Archiving " + obj.source + " to " + zipOutput);
            var output = fs.createWriteStream(zipOutput);
            var archive = archiver("zip");
            output.on("close", function() {
              _logSuccess(
                obj.package.name +
                  "_" +
                  obj.package.version +
                  "_" +
                  obj.package.date +
                  ".zip " +
                  archive.pointer() +
                  " total bytes"
              );
              callback("Done");
            });
            archive.on("error", function(err) {
              _logError(err);
              throw err;
            });
            archive.pipe(output);
            archive.directory(obj.source, "");

            //archive.bulk([{
            //expand: true,
            //cwd: obj.source,
            //src: ['**/*'],
            //dest: ""
            //}]);

            archive.finalize();
          } else {
            callback("Done");
          }
        }
      });
    }
  });
};

module.exports = buildPackage;
